store/tikv/gcworker/gc_worker.go

```diff
@@ -422,6 +423,127 @@ func (w *GCWorker) deleteRanges(ctx goctx.Context, safePoint uint64) error {
+	startKey  []byte
+	endKey    []byte
+	safePoint uint64
+}
+
+type gcResult struct {
+	task           *gcTask
+	successRegions int
+	failedRegions  int
+	err            error
+}
+
+func getNextGCTask(store tikv.Storage, bo *tikv.Backoffer, safePoint uint64, lastKey kv.Key) (*gcTask, error) {
+	loc, err := store.GetRegionCache().LocateKey(bo, lastKey)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	task := &gcTask{
+		startKey:  lastKey,
+		endKey:    loc.EndKey,
+		safePoint: safePoint,
+	}
+	return task, nil
+}
+
+func doGC(ctx goctx.Context, store tikv.Storage, safePoint uint64, identifier string) error {
+	gcWorkerCounter.WithLabelValues("do_gc").Inc()
+
+	err := saveSafePoint(store.GetSafePointKV(), tikv.GcSavedSafePoint, safePoint)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Sleep to wait for all other tidb instances update their safepoint cache.
+	time.Sleep(gcSafePointCacheInterval)
+
+	successRegions := 0
+	failedRegions := 0
+	remained := 0
+	startTime := time.Now()
+	log.Infof("[gc worker] %s start gc, safePoint: %v.", identifier, safePoint)
+	defer func() {
+		log.Infof("[gc worker] %s finish gc, running jobs %v, safePoint: %v, successful regions: %v,"+
+			" failed regions: %v, cost time: %s",
+			identifier, remained, safePoint, successRegions, failedRegions, time.Since(startTime))
+		gcHistogram.WithLabelValues("do_gc").Observe(time.Since(startTime).Seconds())
+	}()
+
+	var key []byte
+	gcResultChan := make(chan *gcResult, gcDefaultGCConcurrency)
+	bo := tikv.NewBackoffer(tikv.GcGetTaskMaxBackoff, goctx.Background())
+	jobCtx, cancel := goctx.WithCancel(ctx)
+	defer cancel()
+	for i := 0; i < gcDefaultGCConcurrency; i++ {
+		task, err := getNextGCTask(store, bo, safePoint, key)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		key = task.endKey
+
+		if task != nil {
+			go doGCTask(jobCtx, store, identifier, gcResultChan, task)
+			remained++
+		}
+
+		if len(key) == 0 {
+			break
+		}
+	}
+
+	ticker := time.NewTicker(gcJobLogTickInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.New("[gc worker] gc job canceled")
+		case <-ticker.C:
+			log.Infof("[gc worker] %s gc in process, running jobs %v, safePoint: %v, successful regions: %v, "+
+				"failed regions: %v, cost time: %s",
+				identifier, remained, safePoint, successRegions, failedRegions, time.Since(startTime))
+		case res := <-gcResultChan:
+			remained--
+
+			successRegions += res.successRegions
+			failedRegions += res.failedRegions
+
+			if len(key) == 0 {
+				if remained == 0 {
+					return nil
+				}
+
+				continue
+			}
+
+			task, err := getNextGCTask(store, bo, safePoint, key)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			key = task.endKey
+
+			if task != nil {
+				go doGCTask(jobCtx, store, identifier, gcResultChan, task)
+				remained++
+			}
+		}
+	}
+}
+
+func doGCTask(ctx goctx.Context, store tikv.Storage, identifier string, resChan chan *gcResult, task *gcTask) {
```

Comment:
This function should wait on a task queue and handle task one by one util receive nil.